Add SyncThru binary sensors

nielstron opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix #48102 by manually defining a Syncthru-state-human dictionary based on what is available in the home assistant const file

Note: this is not yet urgent, as the dictionary is still present in pysycthru 0.7.0, however should be introduced before upgrading pysyncthru to the current state

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #48102
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/17517

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

nielstron committed:
Fix missing dictionary SyncThru-Human translation

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -27,6 +36,16 @@
+    SyncthruState.INVALID: STATE_PROBLEM,
+    SyncthruState.OFFLINE: STATE_OFF,
+    SyncthruState.NORMAL: STATE_OK,
+    SyncthruState.UNKNOWN: STATE_UNKNOWN,
```
We represent unknown state with `None`. It will be translated by the base entity.

nielstron replied:
No it is not intended to be a binary sensor. The issue is that the printer can assume a multitude of states (as can be seen by the number of state constants) but I struggled to find the correct "translation" to a human readable state while allowing compatibility with the HA standards. Is this a reasonable approach? Which constants could be used instead?

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -27,6 +36,16 @@
 DEFAULT_MONITORED_CONDITIONS.extend([f"output_tray_{key}" for key in OUTPUT_TRAYS])
 
+SYNCTHRU_STATE_HUMAN = {
+    SyncthruState.INVALID: STATE_PROBLEM,
```
State problem / state ok are normally binary sensor values. Should this be a binary sensor instead?

nielstron replied:
No it is not intended to be a binary sensor. The issue is that the printer can assume a multitude of states (as can be seen by the number of state constants) but I struggled to find the correct "translation" to a human readable state while allowing compatibility with the HA standards. Is this a reasonable approach? Which constants could be used instead?

MartinHjelmare replied:
If the sensor doesn't use a decide class we don't impose limits on the sensor value, besides the common features like unknown state and unavailable state which we represent in specific ways.

Maybe we don't need a translation of the other states?

nielstron replied:
"Warning" and "error" should definitely be represented by (preferably distinct) values. I realize the sensor might acutally be accurately represented as a binary sensor with OK/NOK (and unavailable/ unknown).

nielstron committed:
Add better translations, introduce seperate sensor

nielstron commented:
I have now split the sensor in two separate sensors. 

One reports whether the device is online and more detailed information through its attributes. This is a breaking change involving the overall state of the main sensor, that has now been moved to its attributes.

The other sensor reports whether the printer is operating correctly or not.

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -173,8 +186,25 @@ async def async_update(self):
-        self._attributes = {"display_text": self.syncthru.device_status_details()}
+        self._state = self.syncthru.is_online()
+        self._attributes = {
+            "detailed_state": SYNCTHRU_STATE_HUMAN[self.syncthru.device_status()],
```
Can we instead create separate sensors for the detailed state and disable that sensor by default?

nielstron replied:
Is disabling via setting "enabled" to False correct? Or is there a different kind of disabling?

MartinHjelmare replied:
We need to use this:
https://developers.home-assistant.io/docs/entity_registry_disabled_by#integrations-setting-default-value-of-disabled_by-for-new-entity-registry-entries

nielstron replied:
I am wondering: is the "async_update" method called for disabled entities? because updates in this sensor are (as of now) the only "real" updates, all other sensors simply rely on this sensor to fetch the new data.

MartinHjelmare replied:
No, the default polling will be turned off for disabled entities.

nielstron replied:
I see. Then the polling should move outside of this entity - maybe into an independent async tracking task?

MartinHjelmare replied:
Use a data coordinator instead:
https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -173,8 +186,25 @@ async def async_update(self):
+        }
+
+
+class SyncThruProblemSensor(SyncThruSensor):
```
If it's a binary sensor it should move to a binary sensor platform.

nielstron replied:
I was unsure whether I could merge the current Syncthru super class with the BinaryEntitty super class and have now defined a binary syncthru super class separately. I hope this fulfils the intention.

MartinHjelmare replied:
We need to create a `binary_sensor.py` module where we create the entities and add them etc.

nielstron replied:
Okay, I split the definitions. Do I have to notify the config_flow setup that there are multiple sensor types for this integration or does it find out by itself?

MartinHjelmare replied:
We should forward the config entry setup to each platform from the init module `async_setup_entry`.

haydee-digitalenachhilfe replied:
done

nielstron replied:
done

nielstron committed:
Make detailed information a separate sensor, disabled per default

nielstron committed:
Untangle sensors, make some binary sensors

nielstron committed:
New fronius version

nielstron committed:
Split setup into binary and normal sensor

nielstron committed:
Fix entity registry per default

nielstron committed:
Fix isort

MartinHjelmare commented:
There are a couple of merge conflicts.

nielstron committed:
Create binary sensors on setup

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -154,7 +179,12 @@ async def async_update(self):
+            "display_text": self.syncthru.device_status_details(),
+        }
+
+    def entity_registry_enabled_default(self) -> bool:
```
```suggestion
    @property
    def entity_registry_enabled_default(self) -> bool:
```

MartinHjelmare commented on homeassistant/components/syncthru/binary_sensor.py:
```diff
@@ -0,0 +1,107 @@
+    def unique_id(self):
+        """Return unique ID for the sensor."""
+        serial = self.syncthru.serial_number()
+        return serial + self._id_suffix if serial else super().unique_id
```
Please use an f-string instead of string concatenation.
```suggestion
        return f"{serial}{self._id_suffix}" if serial else None
```

MartinHjelmare commented on homeassistant/components/syncthru/binary_sensor.py:
```diff
@@ -0,0 +1,107 @@
+    printer = hass.data[DOMAIN][config_entry.entry_id]
+
+    name = config_entry.data[CONF_NAME]
+    devices = [
+        SyncThruOnlineSensor(printer, name),
+        SyncThruProblemSensor(printer, name),
+    ]
+
+    async_add_entities(devices, True)
```
```suggestion
    entities = [
        SyncThruOnlineSensor(printer, name),
        SyncThruProblemSensor(printer, name),
    ]

    async_add_entities(entities, True)
```

MartinHjelmare commented on homeassistant/components/syncthru/binary_sensor.py:
```diff
@@ -0,0 +1,107 @@
+        super().__init__(syncthru, name)
+        self._id_suffix = "_online"
+
+    def is_on(self):
```
This property can move to the parent class.

nielstron committed:
Merge branch 'dev' of github.com:home-assistant/core into patch-missing-syncthru-human

nielstron committed:
Implement changes suggested by MartinHjelmare

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -26,6 +26,26 @@
 DEFAULT_MONITORED_CONDITIONS.extend([f"tray_{key}" for key in TRAYS])
 DEFAULT_MONITORED_CONDITIONS.extend([f"output_tray_{key}" for key in OUTPUT_TRAYS])
 
+SYNCTHRU_STATE_PROBLEM = {
```
This isn't used here.

MartinHjelmare reviewed (changes requested):
Sorry, missed a spot.

nielstron committed:
Remove unused dictionary

MartinHjelmare commented:
Please link a docs PR in this PR description. We want to add the binary sensor platform to `ha_platforms`.

nielstron committed:
Introduce data update coordinator

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -182,21 +168,23 @@ def entity_registry_enabled_default(self) -> bool:
+    @property
+    def extra_state_attributes(self):
+        """Show all data returned for this toner."""
+        return self.syncthru.toner_status().get(self._color, {})
```
It would be good to set the state attribute keys explicitly here so we csn check what they are and their formatting. Can be done in a follow up PR too.

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -210,13 +198,15 @@ def __init__(self, syncthru, name, color):
+    @property
+    def extra_state_attributes(self):
+        """Show all data returned for this drum."""
+        return self.syncthru.drum_status().get(self._color, {})
```
Same here.

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -229,15 +219,18 @@ def __init__(self, syncthru, name, number):
+    @property
+    def extra_state_attributes(self):
+        """Show all data returned for this input tray."""
+        return self.syncthru.input_tray_status().get(self._number, {})
```
And here.

MartinHjelmare commented on homeassistant/components/syncthru/sensor.py:
```diff
@@ -250,12 +243,15 @@ def __init__(self, syncthru, name, number):
+    @property
+    def extra_state_attributes(self):
+        """Show all data returned for this output tray."""
+        return self.syncthru.output_tray_status().get(self._number, {})
```
And here.

MartinHjelmare reviewed (approved):
Nice!

nielstron committed:
Fix formatting issues

nielstron committed:
Merge branch 'dev' of github.com:home-assistant/core into patch-missing-syncthru-human

nielstron committed:
Formatting

MartinHjelmare commented on homeassistant/components/syncthru/binary_sensor.py:
```diff
@@ -0,0 +1,110 @@
+"""Support for Samsung Printers with SyncThru web interface."""
```
Please exclude this module from coverage calculation in `.coveragerc`.

MartinHjelmare commented on homeassistant/components/syncthru/__init__.py:
```diff
@@ -60,12 +75,12 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
 
-def device_identifiers(printer: SyncThru) -> set[tuple[str, ...]]:
+def device_identifiers(printer: SyncThru):
```
Please type the signature completely, including return value, if typing any part of the signature.

nielstron replied:
Ok. I will remove the signature completely since the current setup forbids me to set the correct signature The reason is that `set` and `tuple` (lowercase), cannot be typed, however some tool in the pre-commit chain before that rewrites `Set` and `Tuple` to their lowercase correspondences.

MartinHjelmare replied:
`set` and `tuple` can be used as type annotations.

nielstron committed:
Implement changes requested by MartnHjelmare

remove coverage check for syncthru binary sensor
remove typing of syncthru device identifiers

MartinHjelmare commented on .coveragerc:
```diff
@@ -966,6 +966,7 @@ omit =
     homeassistant/components/switcher_kis/switch.py
     homeassistant/components/switchmate/switch.py
     homeassistant/components/syncthru/sensor.py
+    homeassistant/components/syncthru/binary_sensor.py
```
```suggestion
    homeassistant/components/syncthru/__init__.py
    homeassistant/components/syncthru/binary_sensor.py
    homeassistant/components/syncthru/sensor.py
```

MartinHjelmare committed:
Fix coverage

MartinHjelmare committed:
Revert manifest formatting

MartinHjelmare committed:
Break long string

MartinHjelmare committed:
Fix device identifiers and typing

MartinHjelmare committed:
Revert strings formatting

MartinHjelmare committed:
Remove not needed update

MartinHjelmare reviewed (approved):
Thanks!
